### PR TITLE
[CA 203227] VM shutdown is very slow with many VDIs

### PR DESCRIPTION
**Pull request by mcintyre94**
The reason that VM shutdown was very slow with many VDIs is because the destroy_sr code becomes slow with a large number of VDIs attached : for every VDI it attempts to lock it, then it checks if it has the datapath being removed. In practice this is unnecessary, and on this basis we can drastically improve the performance. This code makes 2 assumptions, which it verifies - raising an exception if they become untrue.

1)  A datapath is only ever on zero or one VDIs. 
2) The VDIs containing a datapath aren't changed by another thread while the destroy_sr call works.

Assumption 2 means that we don't need to lock every VDI to check if it has the datapath, we can just do the check at the beginning - checking without locking is very cheap, it's the lock contention that causes poor performance. Assumption 1 allows further simplification meaning we can work with Options instead of lists. Note that the signature of destroy_sr changes due to assumption 1, it now returns an Option (None or Some exception) instead of a list of exceptions (since it only works on the Option (vdi, vdi_t) - not all of the VDIs. 

Also note that the debug code adds to the runtime complexity of this function (most notably we check every VDI again to verify assumption 2), but this code is no longer the bottleneck and observed shutdown performance wouldn't improve drastically if that was removed. 

**Performance numbers (shutdown VM with 255 VDIs):**
Before: ~3m30s
After: ~1m40s
For comparison, starting this VM takes ~1m50s.

### CODE REVIEW

File: ocaml/xapi/storage_impl.ml [hunk 1]
```diff
@@ -619,24 +622,80 @@ module Wrapper = functor(Impl: Server_impl) -> struct
...
+        if vdi_already_locked
+          then fun f -> f ()
+          else VDI.with_vdi sr vdi in
+
+      (* This is debug code to verify that no more than 1 VDI matched the datapath. We also convert the 0 and 1 cases to an Option which is more natural to work with *)
+      let vdi_to_remove = match vdis_with_dp with
+        | [] -> None
+        | [x] -> Some x
+        | _ -> 
+           raise (Storage_interface.Backend_error (Api_errors.internal_error, [Printf.sprintf "Expected 0 or 1 VDI with datapath, had %d" 1]));
```
Comment (mseri) [inline ocaml/xapi/storage_impl.ml]: Looks like a count to the number of VDIs to display in the %d is missing and there is a 1 instead

Comment (mcintyre94) [inline ocaml/xapi/storage_impl.ml]: Oops, will fix - thanks!

File: ocaml/xapi/storage_impl.ml [hunk 2]
```diff
@@ -619,24 +622,80 @@ module Wrapper = functor(Impl: Server_impl) -> struct
...
+      
+      (* Function to see if a (vdi, vdi_t) matches vdi_ident *)
+      let matches (vdi, vdi_t) = match vdi_ident with
+        | None -> false
+        | Some s -> vdi = s
+      in
+
+      let race_occured  =  match vdis_with_dp with 
+        | [] -> false
+        | [(vdi, vdi_t)] -> not (matches (vdi, vdi_t))
```
Comment (mseri) [inline ocaml/xapi/storage_impl.ml]: Why not `[v] -> not (matches v)` without pattern matching on the tuple?

**Comment by robhoes** [comment]
Please add a signed-off-by line at the end of the commit messages.

File: ocaml/xapi/storage_impl.ml [hunk 3]
```diff
@@ -619,24 +622,80 @@ module Wrapper = functor(Impl: Server_impl) -> struct
...
+        List.map (fun(vdi, vdi_t) -> Printf.sprintf "VDI found with the dp is %s" vdi) vdis_with_dp in
+        raise (Storage_interface.Backend_error (Api_errors.internal_error, message));
+      );      
+
+      failure
 
     let destroy context ~dbg ~dp ~allow_leak =
       info "DP.destroy dbg:%s dp:%s allow_leak:%b" dbg dp allow_leak;
-      let failures = List.fold_left (fun acc (sr, sr_t) -> acc @ (destroy_sr context ~dbg ~dp ~sr ~sr_t ~allow_leak false)) [] (Host.list !Host.host) in
+      let failures = List.map (fun (sr, sr_t) -> destroy_sr context ~dbg ~dp ~sr ~sr_t ~allow_leak false) (Host.list !Host.host)
```
Comment (lindig) [inline ocaml/xapi/storage_impl.ml]: Two small-scale suggestions: The pipeline theme could be extended by starting with
```
Host.list !Host.host
|> List.map (fun (sr, sr_t) -> destroy_sr context ~dbg ~dp ~sr ~sr_t ~allow_leak false)
|> ...
```

`List.filter(function | None -> false | Some x -> true)` can be written as `List.filter ((<>) None)` if that is not too terse.

Comment (mcintyre94) [inline ocaml/xapi/storage_impl.ml]: I like the second one, knew there must be a better way to write that :)

File: ocaml/xapi/storage_impl.ml [hunk 4]
```diff
@@ -692,8 +692,9 @@ module Wrapper = functor(Impl: Server_impl) -> struct
 
     let destroy context ~dbg ~dp ~allow_leak =
       info "DP.destroy dbg:%s dp:%s allow_leak:%b" dbg dp allow_leak;
-      let failures = List.map (fun (sr, sr_t) -> destroy_sr context ~dbg ~dp ~sr ~sr_t ~allow_leak false) (Host.list !Host.host)
-                   |>List.filter(function | None -> false | Some x -> true)
+      let failures = Host.list !Host.host
+		   |>List.map (fun (sr, sr_t) -> destroy_sr context ~dbg ~dp ~sr ~sr_t ~allow_leak false)
+                   |>List.filter((<>) None)
                    |>List.map(function | None -> assert false | Some x -> x) in
```
Comment (mseri) [inline ocaml/xapi/storage_impl.ml]: The `failures` can be shortened using the `List.filter_map` from stdext (https://github.com/xapi-project/stdext/blob/master/lib/listext.mli#L157) as follows:
```
let failures = Host.list !Host.host
                    |> List.fiter_map (fun (sr, sr_t) -> destroy_sr context ~dbg ~dp ~sr ~sr_t ~allow_leak false)
```
the function `filter_map` drops all the None from the mapped list and extracts the content of all the Some

Comment (mcintyre94) [inline ocaml/xapi/storage_impl.ml]: Thanks - will change. I came across the Core implementation, should've checked if you had one too

File: ocaml/xapi/storage_impl.ml (hunk 3 above)
Comment (thomassa) [inline ocaml/xapi/storage_impl.ml]: Or slightly longer but possibly clearer would be `List.filter (fun x -> x<>None)`
(I'm just making an observation, not recommending a change.)

**Comment by mcintyre94** [comment]
Are there any more comments or is this okay to squash ready for merging now?

**Comment by mcintyre94** [comment]
Squashed :)

**Comment by mseri** [comment]
@thomassa @lindig if you are happy with this, we can merge

**Comment by thomassa** [comment]
In it goes then.

**Pull request merged**